benchmark/src/main/java/org/neo4j/prof/FlameGraph.java

```diff
@@ -0,0 +1,336 @@
+package org.neo4j.prof;
+
+import joptsimple.ArgumentAcceptingOptionSpec;
+import joptsimple.OptionParser;
+import joptsimple.OptionSet;
+import joptsimple.OptionSpec;
+import org.neo4j.graphalgo.helper.ldbc.LdbcDownloader;
+import org.openjdk.jmh.infra.BenchmarkParams;
+import org.openjdk.jmh.infra.IterationParams;
+import org.openjdk.jmh.profile.ExternalProfiler;
+import org.openjdk.jmh.profile.InternalProfiler;
+import org.openjdk.jmh.profile.ProfilerException;
+import org.openjdk.jmh.results.BenchmarkResult;
+import org.openjdk.jmh.results.IterationResult;
+import org.openjdk.jmh.results.Result;
+import org.openjdk.jmh.runner.IterationType;
+import org.openjdk.jmh.util.Utils;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.OptionalLong;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public final class FlameGraph implements InternalProfiler, ExternalProfiler {
+
+    private static final String ASYNC_PROFILER_DIR = "ASYNC_PROFILER_DIR";
+    private static final String DEFAULT_EVENT = "cpu";
+
+    private final String event;
+    private final Path asyncProfilerDir;
+    private final boolean threads;
+    private final OptionalLong frameBufferSize;
+    private final OptionalLong interval;
+    private final Collection<Path> generated = new ArrayList<>();
+    private Path outputDir;
+    private boolean started;
+    private int measurementIterationCount;
+
+    public FlameGraph(final String initLine) throws ProfilerException {
+        Path asyncDir = null;
+        String env = System.getenv(ASYNC_PROFILER_DIR);
+        if (env != null) {
+            asyncDir = Paths.get(env).toAbsolutePath();
+        }
+
+        OptionParser parser = new OptionParser();
+
+        OptionSpec<String> outputDir = parser
+                .accepts("dir", "Output directory")
+                .withRequiredArg()
+                .describedAs("directory")
+                .ofType(String.class);
+
+        ArgumentAcceptingOptionSpec<String> asyncProfiler = parser
+                .accepts(
+                        "asyncProfiler",
+                        "Location of https://github.com/jvm-profiling-tools/async-profiler unless procived by $" + ASYNC_PROFILER_DIR)
+                .withRequiredArg()
+                .ofType(String.class)
+                .describedAs("directory");
+        if (asyncDir == null) {
+            asyncProfiler.required();
+        } else {
+            asyncProfiler.defaultsTo(asyncDir.toString());
+        }
+
+        OptionSpec<String> event = parser
```

Comment:
We already have some of those, others aren't applicable anymore:
- `debugNonSafepoints` is forced to true, as is [suggested by the profiler](https://github.com/jvm-profiling-tools/async-profiler/blob/master/README.md#restrictionslimitations) (last bullet point)
- `verbose` is, as far as I can see, used to pipe the profiler output to the benchmark output stream, in case one doesn't want flame graphs and just the default async-profiler summary. Since we only want flame graphs, we don't need this option
- `flameGraphDir` not necessary, as this is directly done by async-profiler
- `jfr` is not the target responsibility for this profiler, we already have a JFR profiler

Some, we could make use of:

- `flameGraphOpts`: async-profiler has its own set of options to modify the flame graphs
- `flameGraphDirection` just a special case to see if `flameGraphOpts=--reverse` should be applied 
- `simpleName` modifies the intermediate result before calling the flame graph script on it, async-profiler has an equivalent option `-s` 

I can add a `async-profiler-opts` option to pass any additional options to the profiler call, which would then cover the last 3 options.

Reply:
Perfect, an argument for arguments is what I had in mind, too :)